includes/class-wcs-att-scheme.php

```diff
@@ -81,11 +83,50 @@ public function __construct( $args ) {
 			if ( 'inherit' === $this->data[ 'pricing_mode' ] ) {
 				$this->data[ 'discount' ] = isset( $args[ 'data' ][ 'subscription_discount' ] ) ? wc_format_decimal( $args[ 'data' ][ 'subscription_discount' ] ) : '';
 			}
+
+			$this->data[ 'sync_date' ] = 0;
+
+			if ( isset( $args[ 'data' ][ 'subscription_payment_sync_date' ] ) ) {
+
+				if ( is_array( $args[ 'data' ][ 'subscription_payment_sync_date' ] ) && isset( $args[ 'data' ][ 'subscription_payment_sync_date' ][ 'day' ] ) && isset( $args[ 'data' ][ 'subscription_payment_sync_date' ][ 'month' ] ) ) {
+
+					$this->data[ 'sync_date' ] = array(
+						'day'   => $args[ 'data' ][ 'subscription_payment_sync_date' ][ 'day' ],
+						'month' => $args[ 'data' ][ 'subscription_payment_sync_date' ][ 'month' ]
+					);
+
+				} else {
+					$this->data[ 'sync_date' ] = absint( $args[ 'data' ][ 'subscription_payment_sync_date' ] );
+				}
+			}
 		}
 
 		$this->data[ 'context' ] = isset( $args[ 'context' ] ) ? strval( $args[ 'context' ] ) : 'product';
 
 		$this->key = implode( '_', array_filter( array( $this->data[ 'interval' ], $this->data[ 'period' ], $this->data[ 'length' ] ) ) );
+
+		$this->data[ 'key' ] = $this->data[ 'id' ] = $this->key;
+
+		/*
+		 * Syncing & Proration.
+		 */
+
+		$this->data[ 'is_synced' ]   = false;
+		$this->data[ 'is_prorated' ] = false;
+
+		if ( class_exists( 'WC_Subscriptions_Synchroniser' ) ) {
+
+			$dummy_product = new WC_Product( 0 );
+
+			WCS_ATT_Product_Schemes::set_subscription_schemes( $dummy_product, array(
+				$this->get_key() => $this
+			) );
+
+			WCS_ATT_Product_Schemes::set_subscription_scheme( $dummy_product, $this->get_key() );
+
+			$this->data[ 'is_prorated' ] = WC_Subscriptions_Synchroniser::is_product_prorated( $dummy_product );
+			$this->data[ 'is_synced' ]   = WC_Subscriptions_Synchroniser::is_product_synced( $dummy_product );
```

Comment:
Nice analysis and agreed. See https://github.com/Prospress/woocommerce-subscribe-all-the-things/pull/278#issuecomment-381312327

Needs fixing!

Reply:
You are right about `wcs-att-single-add-to-cart.js` -- but there we already have a product context, so the `is_prorated` flag can be populated with that specific product in mind.

Reply:
> `$this->is_prorated()` is only used by `WCS_ATT_Manage_Add_Product::options_template()`

There's also `WCS_ATT_Scheme::get_data()`. Also remember that in order to check whether a product needs to be prorated, we must temporarily apply that scheme on the object, which is standard practice in the `WCS_ATT_Product_Schemes` API.

Generally speaking, proration is a result of:

* the global proration/sync settings;
* a scheme's sync settings; and
* the subject (product).

The result being a combination of the product + a scheme, I'm inclined to introduce a `WCS_ATT_Product_Schemes::is_subscription_scheme_prorated( $product, $scheme_key )` method, or a `WCS_ATT_Product::is_prorated( $product, $scheme )`, which would be a more generic way to test if a product would be prorated assuming a scheme is activated on it, without necessarily requiring that the scheme being tested is one of those tied to the object.

It's a tough call.

Reply:
> or a `WCS_ATT_Product::is_prorated( $product, $scheme )`, which would be a more generic way to test if a product would be prorated assuming a scheme is activated on it, without necessarily requiring that the scheme being tested is one of those tied to the object.

That gets my vote 👍 (although maybe also include something about it being related to _sync_ proration to avoid any confusion with _switch_ proration).

To me, that is more correct because it's checking if a _product_ (+ scheme) is prorated, which is really what proration is, not if a _scheme_ (on a given product) is prorated.

Reply:
+1

@thenbrent could you perhaps give some details on sync vs switch proration? Are there different methods for those contexts in WCS?

Reply:
Looking at the WCS switching class, I get what you mean now -- there are dedicated settings that control whether proration should happen when switching, vs proration for the 1st payment only.

In this case, we have the following alternatives:

* Add `is_prorated( $product, $scheme, $context )` to the `WCS_ATT_Product` class, but then we'd need to handle "1st payment" vs "switch" context separately, and in "switch" context there would need to be something extra depending on whether we're checking price, sign up fee, or length proration. This could be an extra argument, or it could be embedded in the `context`, e.g. `switch:price`, `switch:sign-up-fee`, etc.
* Add `is_prorated` to the `WCS_ATT_Sync` class, and (if needed later) add another method in the `WCS_ATT_Switch` class.

I see that WCS currently doesn't have a dedicated method to check for "switch" context proration.

Reply:
Will proceed with the second option and cc you on the commit after doing some testing to ensure nothing is broken.